src/components/astra-editor/index.ts

```diff
@@ -0,0 +1,314 @@
+import { getPredefineTheme } from './theme'
+
+@customElement('astra-editor')
+export class AstraEditor extends LitElement {
+  protected editor?: EditorView
+  protected extensions: { name: string; comp: Compartment; ext: Extension }[] = []
+  protected containerRef: Ref<HTMLDivElement> = createRef()
+  protected previousValue: string = ''
+
+  @property() color: string = 'light'
+  @property({ type: 'Boolean' }) wrap: boolean = false
+  @property() theme: string = 'moondust'
+  @property() placeholder: string = ''
+  @property() value: string = ''
+
+  static styles = css`
+    .cm-tooltip-autocomplete ul::-webkit-scrollbar,
+    .cm-scroller::-webkit-scrollbar {
+      width: 10px;
+      height: 10px;
+    }
+
+    .cm-scroller::-webkit-scrollbar-thumb {
+      border-radius: 5px;
+    }
+
+    .cm-tooltip-autocomplete ul::-webkit-scrollbar-thumb,
+    .cm-scroller::-webkit-scrollbar-thumb {
+      background: rgba(0, 0, 0, 0.3);
+    }
+
+    .dark .cm-tooltip-autocomplete ul::-webkit-scrollbar-thumb,
+    .dark .cm-scroller::-webkit-scrollbar-thumb {
+      background: rgba(255, 255, 255, 0.3);
+    }
+
+    .cm-tooltip-autocomplete > ul > li {
+      display: flex;
+    }
+
+    .cm-tooltip-autocomplete > ul > li > .cm-completionIcon {
+      width: 1em !important;
+      display: flex;
+      align-self: center;
+      justify-content: center;
+    }
+
+    .cm-tooltip-autocomplete .cm-completionLabel {
+      flex-grow: 1;
+    }
+
+    .cm-tooltip-autocomplete .cm-completionDetail {
+      padding-left: 15px;
+    }
+
+    .cm-completionIcon-property::after {
+      content: '\\61' !important;
+      font-family: 'outerbase-icon' !important;
+    }
+
+    .cm-completionIcon-type::after {
+      content: '\\62' !important;
+      font-family: 'outerbase-icon' !important;
+    }
+
+    .cm-completionIcon-function::after,
+    .cm-completionIcon-method::after,
+    .cm-completionIcon-variable::after,
+    .cm-completionIcon-namespace::after,
+    .cm-completionIcon-interface::after {
+      content: '⚡' !important;
+    }
+  `
+
+  constructor() {
+    super()
+
+    // We will use icon font instead of SVG because CodeMirror rely
+    // on :after content. It is easily to change the color of the icon
+    // when we toggle dark/light mode.
+    //
+    // We append it to document instead of shadowRoot because browser
+    // does not support loading font dynamically inside shadowRoot
+    //
+    // To construct the font, you can do it on
+    // https://icomoon.io/app/
+    // \62 => https://phosphoricons.com/?q=%22column%22
+    // \61 => https://phosphoricons.com/?q=%22table%22
+    if (!document.getElementById('codemirror-custom-font')) {
+      const head = document.head || document.getElementsByTagName('head')[0]
+      const iconFontStyle = document.createElement('style')
+      iconFontStyle.id = 'codemirror-custom-font'
+      head.appendChild(iconFontStyle)
+
+      iconFontStyle.innerHTML = `
+        @font-face {
```

Comment:
can this be declared in the `styles` section above now too (maybe in a [`:host { }`](https://developer.mozilla.org/en-US/docs/Web/CSS/:host_function) section), or is there something about it being in the `<head />` tag that's necessary for it to work right?